### PR TITLE
broadcasting docs

### PR DESCRIPTION
**Pull request by GeorgeNash**
Rather than share around the doc on broadcasting I thought it'd be better just to place it on the Tensor docs.

A few examples of basic broadcasting mechanics.

### CODE REVIEW

File: docs/current_docs/tensors.md [hunk 1]
```diff
@@ -142,6 +142,131 @@ GaussianVertex vertex = new GaussianVertex(shape, mu, 0);
...
+* shape (the length in each dimension). 
+
+Tensors can be added, subtracted, multiplied and more with each other. 
+It wouldn’t be very useful if you could only operate on tensors of the same shape, fortunately you can perform broadcasting.  
+
+### Tensor and Scalar
+
+Let’s start with a simple example, multiplying each element inside a matrix by a constant.
+
+Let’s define A as a tensor of shape [2, 2]. It’s therefore a 2x2 matrix of rank 2.
```
Comment (alexsparrow) [inline docs/current_docs/tensors.md]: Maybe put backticks around the shapes here to match above and below.

File: docs/current_docs/tensors.md [hunk 2]
```diff
@@ -142,6 +142,131 @@ GaussianVertex vertex = new GaussianVertex(shape, mu, 0);
...
+    30., 40.
+}, 2, 2); //shape [2, 2]
+
+DoubleTensor B = DoubleTensor.scalar(5.); //shape []
+
+DoubleTensor C = A.times(B); //shape [2, 2]
+```
+
+We can do operations as normal between the two, even though they have different shapes. 
+In the case of a tensor and a scalar, the resulting tensor will always have the shape of the initial tensor.
```
Comment (migwellian) [inline docs/current_docs/tensors.md]: I think you can drop "In the case of a tensor and a scalar" - it aids the clarity and we know you are only talking about this case.

File: docs/current_docs/tensors.md [hunk 3]
```diff
@@ -142,6 +142,131 @@ GaussianVertex vertex = new GaussianVertex(shape, mu, 0);
...
+
+
+### Tensor and Tensor
+
+You will want to operate on tensors of different sizes.
```
Comment (migwellian) [inline docs/current_docs/tensors.md]:
```suggestion
Sometimes you will want to operate on tensors of different sizes. 
```

File: docs/current_docs/tensors.md [hunk 4]
```diff
@@ -142,6 +142,131 @@ GaussianVertex vertex = new GaussianVertex(shape, mu, 0);
...
+#### Subset of shape
+
+*Valid Example*
+
+Operating on a `[2, 2, 2]` and a `[2, 2]`
+
+```java
+DoubleTensor A = DoubleTensor.create(new double[]{
+    10., 20., 30., 40.,
+    50., 60., 70., 80.
```
Comment (migwellian) [inline docs/current_docs/tensors.md]: Could you format this on the page as a [2,2,2]? i.e.
```
10 20
30 40

50 60
70 80
```

File: docs/current_docs/tensors.md [hunk 5]
```diff
@@ -142,6 +142,131 @@ GaussianVertex vertex = new GaussianVertex(shape, mu, 0);
...
+}, 2, 2, 2);  //shape [2, 2, 2]
+
+DoubleTensor B = DoubleTensor.arange(1, 8).reshape(2, 2);  //shape [2, 2]
```
Comment (migwellian) [inline docs/current_docs/tensors.md]: Shouldn't this be `arange(0, 4)`?

File: docs/current_docs/tensors.md [hunk 6]
```diff
@@ -142,6 +142,131 @@ GaussianVertex vertex = new GaussianVertex(shape, mu, 0);
...
+
+DoubleTensor C = A.times(B); //shape [2, 2, 2]
+DoubleTensor D = A.plus(B); //shape[2, 2, 2]
```
Comment (migwellian) [inline docs/current_docs/tensors.md]: For clarity of what the operation is doing, you should print out the results

File: docs/current_docs/tensors.md [hunk 7]
```diff
@@ -142,6 +142,131 @@ GaussianVertex vertex = new GaussianVertex(shape, mu, 0);
...
+```
+
+*Invalid Example*
+
+Operating on a `[2, 2, 3]` and a `[2, 2]`
+
+```java
+DoubleTensor A = DoubleTensor.create(new double[]{
+    10., 20., 30., 40., 50., 60.,
+    70., 80., 90., 100., 110., 120.
```
Comment (migwellian) [inline docs/current_docs/tensors.md]: format on the page as [2, 2, 3]

File: docs/current_docs/tensors.md [hunk 8]
```diff
@@ -142,6 +142,131 @@ GaussianVertex vertex = new GaussianVertex(shape, mu, 0);
...
+}, 2, 2, 3);  //shape [2, 2, 3]
+
+DoubleTensor B = DoubleTensor.arange(1, 8).reshape(2, 2);  //shape [2, 2]
```
Comment (migwellian) [inline docs/current_docs/tensors.md]: arange(0, 4) ?

File: docs/current_docs/tensors.md (hunk 5 above)
Comment (GeorgeNash) [reply docs/current_docs/tensors.md]: Good spot!! Copy pasting error

File: docs/current_docs/tensors.md (hunk 6 above)
Comment (GeorgeNash) [reply docs/current_docs/tensors.md]: This section is really just focusing on which shapes are broadcastable. It's a good point though - I've added the results of the operation to the complete example just below this one.

**Pull request merged**